### DISCUSSION TITLE
Voting mechanism (Technical discussions about how the process of voting could be implemented)

### DISCUSSION THREAD

**Discussion by Phyloiota** [iota-community/Community-Governance#5] [Ideas]
We need to find a way to use the IOTA protocol for a fair and immutable voting system. What will be technically doable with the current protocol or future implementations?

**Comment by Phyloiota** [comment]
It seems the initial vote and decision if tokens will be used in governance, or burned might be purely based on token holdings, as this is the only possible solution that we might be able to implement in the near future (2 months,). So i would like to bring a few ideas up how this can be achieved with the protocol in the current state or some small modifications:

Technical requirements and possibilities for the voting mechanism 
“Token balance under control of a voter” = Meaning 1 IOTA = 1 Vote

Most important in my opinion is that it must be an absolute ungamable solution: 
- no one should be able to find a technical solution to generate more votes then the balance that sits on the addresses of the seeds he(or she ofc) controls.
- The technical solution must guarantee that one seed with its balance can only 1 time deliver a vote and no multiple votings from the same seeds / addresses are possible
- The voting must happen “on tangle” with the highest possible security
- The “counting” of the votes needs to happen on tangle and needs to be verifiable by everyone and happen absolute transparent and open source
- All code, tools, solutions used to generate or count the votes must be open source code published on github and should be part of the IOTA protocol or developed for the IOTA protocol.
- Use of any closed source or 3rd party tools needs to be avoided

How to “make a vote?”

So we need to find a way to count how many IOTA tokens vote for option A (treasury) and option B (burning)
I see in the moment 2 possible ways of using the current protocol “as is” and 1 way using Level 1 digital assets (needs to be implemented for the vote) - as i am not 100% sure and not a dev i would like to ask our Specialist and IOTA devs that are more deep in the protocol to comment and verify or offer ways to adjust this ideas:

1.) Voting transaction over voting nodes
A user can verify his ownership over a certain amount of tokens by issuing a valid transaction from the address where those tokens are located. In the current UTXO Ledger i don't see a practical way to make a transaction with a very low amount (1 Mi) to proof this, as after this transaction the balance will be moved to a new address and i believe it is difficult to create a tool that would track all those addresses before and after the voting transaction so my idea is this:

Every voter sends a transaction containing the full balance under control of the owner to another address under his control (internal wallet to wallet transfer in Firefly) and uses a special setup of voting nodes for this transaction. One Node (or quorum) for option A, one Node (or quorum) for option B. This nodes accept only one translation from a seed, further transactions from the same seed will be not allowed (I am not sure how this can be technically guaranteed, but i can think of ways to do this tracking the public key signatures of those transactions)
If a voter has multiple profiles (seeds) he can initiate one vote from every profile (seed), but can never vote with more tokens than he totally owns with this solution.

Counting of votes happens in the nodes. Let's say we have a voting period of 24 hours (or a week, …) 
At the start of this period the nodes for Option A and B go online, users decide in Firefly with manual node selection over which node their voting transaction should be processed.(recommended DNS Addresses like https://IOTA_voting_node_optionA.org…) The voter issues the transaction (internal Wallet transfer) with the full balance (or as much balance they want to put into the vote) of his seed.
Node group A receives this transaction to process. A code reads out “transferred balance” and “public key signature” and stores it in a database. A function gets triggered that this public key signature will not be accepted anymore by the nodes.
At the end of the voting period the database of the node gets a snapshot and the nodes go offline. The community can validate those snapshots similar as it happened in the migration to proof that nothing got cheated. The readout of the snapshot in total token balance transferred over the node is the voting count for Option A or B.

Difficulties of course - who runs the Nodes? A group of community nodes with this special modified software? The IF? Is 1 node for option A and 1 node for option B okay (and even possible?) Who oversees the counting?.....



2.) Voting transaction using 0 value or 1 i tx on voting addresses

The voter sends a data transaction or a 1 i value transaction from his balance address to a voting address that represents choice A and another address represents choice B.. This voting address is created at the beginning of the voting period and will be snapshotted at the end of the voting period. 
Now it needs to be analysed from which addresses transactions have arrived on the voting address and the balances of those addresses need to be read out. Accumulation of the global address balances gives total voted token balance for Option A / Option B.
Problem is the UTXO model that will always move all tokens away from the “starting address” to fresh addresses (outputs). So we would need to create a global snapshot of all addresses at the beginning of the vote. The voters must make sure that after this snapshot, at the beginning of the vote, the first transaction they do needs to be the voting transaction. 
If a voter out of confusion or other reasons does first another transaction before he does the voting transaction this voting transaction will not count as the address from where the tx came from was not part of the global snapshot and therefore cannot be counted.
Using a pure Data transaction might offer the solution, but i dont know if the data transactions are coming from the balance addresses (seeing that nodes can spam transactions without having a balance address)
And if it is possible to send Data transactions from a balance address to another address does it also create new UTXO outputs? So it would not offer an advantage at all…?

3. Voting using unique one time voting tokens (digital assets L1 tokens)

A Function needs to be implemented in Firefly that allows the following:
Create a token (NFT, L1 asset)  that carries the information how many Iota the profil that created the token owns (read balance of seed addresses and stores this information) This NFT can be sent by the voter to a voting address. The balance carried as information engraved in all deposited NFTs on the voting address counts as votes for Option A or Option B.
Things that need to be guaranteed:
Only one token can be created for this specific vote by a seed (voting ID. public key signature,...something like that) 
The voting address needs to be able to read out the information deposited on the NFT
After the vote the tokens would be burned (uncolored) on the voting address and could be returned to the voter address.


This option requires minimum the finished digital asset framework and maybe also ISCP. so i don't know if it is achievable for the current first vote.


Those are the options I see at the moment for a pure 1 vote per 1 IOTA voting mechanism.

I hope to get feedback from our brilliant devs and mathematicians that are more deep in the details of the protocol than I am.
Surely I have some flaws in my technical theories, but I would hope we can come to a usable and bulletproof secure solution using the IOTA protocol.

**Reply by datahomelab** [reply]
Great ideas, and it would be nice to know which one is easier to implement in the short term. Regarding 1i = 1vote, there is still the matter on capping the amount of IOTA, and how to avoid funds split to enable more voting power. Any thought on how to implement this?

**Reply by dudeamir** [reply]
Regarding 3rd mechanism:

is it possible that same number of voting tokens created as the existing iota tokens, then distributed to each address based on its balance (I assume its possible to see the balance by checking public address). So each person wishing to vote, sends their public address to the voting address. Then gets back voting tokens. It has an specific time to send out those voting tokens with a message back to the voting address. (This way there is no way to do double voting)

In regard to UI, in my opinion, non of technicalities of implementation should be a hurdle for IOTA average holder to participate, therefore Firefly UI should do all the job. By clicking on "participate" automatically broadcast its public address as a voter, receiving the tokens. Then by clicking on "Burn" or "Treasury" automatically Firefly does all the transactions with the right message payload.

**Reply by Phyloiota** [reply]
@dudeamir  I think for this functionality it needs smart contracts to automate the creation and distribution of the tokens back to the "owner" addresses.  As this will be  Level 2 tokens created "out of thin air" it can only happen inside the ISCP functionality. So i think we will not be able to do this in our initial vote. For later its for sure a great idea. I think something without smartcontract functionality will maybe possible to achieve, so this is why i suggested the Level 1 token. It is as far as i see possible to do this already with an upgrade of the current software to enable the creation of such an asset. It will be just a new type of payload that it carries on the standard token and tx model.

**Reply by Phyloiota** [reply]
@datahomelab that is another question for another discussion. Of course any cap can be implemented in the counting process, but to decide if and where and how to cap is a more ethical and political question than a technical i think...

**Reply by datahomelab** [reply]
Technical aspect would be how to avoid the split.

**Reply by Phyloiota** [reply]
> Technical aspect would be how to avoid the split.

yeah thats really a thing to think about... i dont see a solution yet that wont lead to discussions and favors subjective decisions... and so could be used to question the validation of the whole voting process. We should try to avoid  anything that smells unfair...monitoring large addresses and movements is possible, but how to distinguish that those moves will be used to gain extra votes through splitting or are real transactions...? If you say tokens need to sit  on addresses a certain time you limit a lot uf users in their freedom of use of the tokens and end up with locked staking pools ---> but without staking rewards.... So only those that would look for a way to influence the vote for personal reasons or profit or revenge would invest the power and time to do that... Everything will lead to a situation where the final result needs to be defeated against suspected fraud and manipulation from different sides... I am very much a fan of limiting the power of certain big player parties, but only if it is a clean, clear and logical objective solution that doesnt have any "smell" of favoring one or another party or opinion on it...

**Reply by datahomelab** [reply]
Agree 100%. I've been racking my brain to think how to solve this one for a couple of days. You explained it very clearly.

**Comment by Phyloiota** [comment]
Whatever solutions is chosen in the end i think we really need to form a kind of Beta Test environment on the Chrysalis-Testnet or a dedicated Testnet and do a lot of Test - votings to see what would be the pros and cons of the different solutions....

**Reply by dudeamir** [reply]
Thats the reason i think to have a dummy public vote prior to the Vote in June can be beneficial to test it not only on a testnet, but also seeing the result out of the real holders of the tokens. Also lots of people would already move their tokens to Firefly from exchanges to be able to do the dummy vote if its interesting enough for them.

**Comment by sdellava** [comment]
If we want avoid any possible discussion about the relevance of the vote or its possible manipulation, thus the correctness of IOTA Foundation's consequent actions, we need to solve a question: how do we define community membership?

The answer may lead to a vote with greater or lesser legal significance, but as far as I know, the amount of IOTA held by a person can only be considered proof of membership if iota were a security token. 

IOTAs would give each member of the community the right to vote in proportion to the amount held, exactly as the amount of stock held gives greater weight to the vote at a company's board meeting. 

I believe that no owner of IOTA tokens wants to be certified that IOTA is Security Token.   

So, I suggest to solve the "membership issue" before to define the voting system.

Then, we can set some additional requirements that can lead to a perfectly legal compliance of the voting procedure:
- The identity of the voter must remain anonymous
- each member of the community can only vote once.
- there must be deadline after which a vote is no longer accepted
- the progressive count of the vote must remain secret until the deadline in order not to influence the behavior of the voters.

**Reply by rophiroth** [reply]
Maybe a more liquid approach can be better.

Changes would be like:

- Identity remain anon by default for normal voters, but public for delegates.
- No deadline, your vote keep save... and if not.. you just fix it once again.
- public count to consider others.

**Reply by ghost** [reply]
this is the crucial topic i think, defining community membership/ validity of the vote

1. simply owning some iota to vote seems insufficient: a wallet can be split up into 50 wallets and one person can vote 50 times

2. then you can consider a weighted vote scenario: you vote with the quantity of iota you possess. then it doesn't matter how many times you vote, your vote will still have the same weight: your total iota ownership amount. but then the problem becomes one of a few or even one whale taking the entire vote with their massive supply dwarfing the community

3. perhaps an upper and lower bound of quantity of iota to vote. but then we're back to the problem in #1, especially for whales

i can't conceive of any easy solution here

**Reply by ghost** [reply]
`The identity of the voter must remain anonymous`

perhaps a DID solution?

such that someone must possess an identity to vote. but the identity does not need to be made public. merely that they have established an identity

perhaps the process for establishing the identity could weed out sybil attacks: one person controlling many sock puppet votes. somehow

**Comment by sascha1337** [comment]
![Screenshot 2021-05-11 at 16 57 30](https://user-images.githubusercontent.com/12372821/117788455-fff58700-b279-11eb-9093-04860025aaf5.png)
Just leaving that information here, to be discussed ...

**Comment by Phyloiota** [comment]
Just putting the conversation that Stefano and i had in Discord up here as i think he brought a very good point forward: Node ID
Phylo [Community DAO - lets go!] — heute um 16:29 Uhr
Thanks a lot! Very valuable input Stefano... a huge load of complexity is involved here ... did absolutely not have this security issue in my thoughts.... Well the thing is i think... this vote will not lead the IF to do anything, it will be an action of the Node operators to implement a software that follows the outcome of the vote and do a ledger update related to this voting decision... So thats what IF is doing - it will try to keep even the smallest toe out of this whole process... 
The security token question is then the same for all the other projects also where your stake gives you an equivalent voting right...
So its on the community to develop a voting system and do a first decision if those funds should disappear or stay on the ledger and be transfered in the control of "an entity formed by community members"
My point was really more if a vote with the IOTA stake can be technically achieved. I am definetly not a big fan of this pure vote per stake mechanism as i value community membership and participation in the project much higher, i just dont see any doable solution of measuring this and give 1 vote per 1 person...
Stefano Della Valle — heute um 16:32 Uhr
Well, if the "action" is in charge of the node operators, we have the rule to define the community membership: a community member is a node operator.

If we set this, we can also say that the voter identity is the node identity.
And then we can say that we already have a very efficient tool to manage network vote: FPC !
This approach let the iota ownership totally apart form the vote, avoiding any compliance issue with the financial law and rules
Phylo [Community DAO - lets go!] — heute um 16:35 Uhr
yeah Mana is definetly the solution for most of that also, but all of this is not implemented. And if i have an interest of gaining more votes i setup 1000 nodes on hetzner... thats a very disputable metric...
Phylo [Community DAO - lets go!] — heute um 16:35 Uhr
this would be the big advantage of this, yes
So if we would be forced to do a vote end of June, how would we achieve a solution?
Stefano Della Valle — heute um 16:39 Uhr
Current network already provide the node identity. We can leverage on that information. 

It is true that one can set up a number of nodes easily, but if that is the membership requirement that's a legal behaviour, even because both kind of voter can do the same, so at the end is not demonstrable that it was an abuse.
Phylo [Community DAO - lets go!] — heute um 16:41 Uhr
okay lets say we use the node id. How do you vote with your node? How do you express your opinion of Option A or B? And how can this vote be unique per Node ID and ungamable?
Stefano Della Valle — heute um 16:43 Uhr
run a plugin could be the easier way submit the vote.
the tech procedure can be based on a simple signature of a message based of the node a private key.
so, in advance, the plug-in need to publish his node identity that will allow to prove the signature.
I guess the fantastic IOTA Identity.rs library can be used to this task
- the plug-in create the node ID and publish the DID Document including his public key
- when the vote is open, the plug in sign e publish a message with yes or no
- the vote include the node DID that allow to resolve the DID document containing the public key
Then we need a tool that everyone can use to count and verify the votes
Phylo [Community DAO - lets go!] — heute um 16:49 Uhr
Excellent - now we are coming forward! That is a solution that is in the right direction! Very much like that idea already...! This is why i came up here with this question.... we need technical solutions to make a vote possible first... and then the "politics" can start!
And so it would also be possible to use both - Node operators gain voting power, but also it would be possible to give Token holders a vote --- maybe give 50% of voting power to NODE ID and 50% voting power to token holders or something like that..... because it will be a big group around that does not run a node, but still wants to vote.... Thank you for that input!
Stefano Della Valle — heute um 16:55 Uhr
That was why I was thinking about FPC: MANA  comes from IOTA so a node can gain voting power thanks to the MAN pledging from  token holder ...
May be we can setup a sort of FPC light for these kind of vote
Phylo [Community DAO - lets go!] — heute um 16:56 Uhr
Yeah Bas and i had Mana also directly in mind - if we would come to a governance solution - then we have time to develop something amazing around Mana and FPC Voting - thats then for the real running mechanism of a DAO.
we would just first get a general decision if those funds will be available for a DAO or if the will disappear und be burned ... then we wont need to worry about all that voting things...

**Reply by Phyloiota** [reply]
Phylo [Community DAO - lets go!] — heute um 18:41 Uhr
Regarding Node ID voting the possible downsides i see:
A plugin for the Hornet Software could be a security risk and must be extremely good developed and tested
It must ensure that Change of Node ID is permitted... So i think to make that more difficult instead of only Node id maybe Node IP address also could be implemented in a DID verification...
And still nothing would stop actors to setup hundreds of Nodes if there is a certain financial (or emotional) incentive involved to influence voting in one way. So we still dont come to a 1 person 1 vote solution with that.
And so as it has some obvious attack vectors those will be used after the vote to discuss and dispute the outcome ... esp. if we get a very close result...

**Reply by basvansambeek** [reply]
To me the *one person, on vote* idea only works if we have an on-tangle Digital Identity connected to your real world identity in a unique and ungameable way.

I don’t see that happening anytime soon, and even then the discussion of what constitutes “community” is still open. Are people hired on fiver part of it? Is my family? Is my extended family, if I set it up for them? But it’s not just excluding certain votes. Can companies vote? Or only the people in it? Can everyone in the company vote, just because they’re interested? Or because they get cake from the one person who is shilling IOTA? Or what if a competing crypto community suddenly decides they want to vote. Because they hate IOTA’s guts? Or just because Boaty McBoatface?

To me ID solutions are out, at least until SSI on IOTA is mature, and maybe not even then. Which I find sad, but I can see it becoming a fair vote anytime soon. Both technically and narrative wise.

**Reply by sdellava** [reply]
> To me the _one person, on vote_ idea only works if we have an on-tangle Digital Identity connected to your real world identity in a unique and ungameable way.
> 
> I don’t see that happening anytime soon, and even then the discussion of what constitutes “community” is still open. Are people hired on fiver part of it? Is my family? Is my extended family, if I set it up for them? But it’s not just excluding certain votes. Can companies vote? Or only the people in it? Can everyone in the company vote, just because they’re interested? Or because they get cake from the one person who is shilling IOTA? Or what if a competing crypto community suddenly decides they want to vote. Because they hate IOTA’s guts? Or just because Boaty McBoatface?
> 
> To me ID solutions are out, at least until SSI on IOTA is mature, and maybe not even then. Which I find sad, but I can see it becoming a fair vote anytime soon. Both technically and narrative wise.

I agree with you about the "personal ID" problem, but to count the votes need a way to identify the voters and validate their options. So the problem here is philosophical: the community is really voting or we just need a way to avoid complains about action that will be take later as result of a moral suasion from huge influencers? 

I guess need a real vote, so we need to grant the voter freedom to vote and to partecipate to the community. 

If this is the magic rule, there is no a real side effect if someone decide to activate a node just to partecipate to the vote and a minute later turn it down. He is free to become a member of the community and there is no additional requirement other than run the node.

In IOTA 2.0 we will have an additional constraint: a node can vote if it is an "active" member of the community. Being an active member means own lot of MANA, thus move IOTA.

I guess there is no time to implement MANA based voting system in the current network (but I hope to be wrong), so the only tool we have is the node id.

If many will take the opportunity to run a node just to partecipate to this pool, I guess it will be simply a good opportunity for the IOTA community to enlarge quickly and involve new adopters.

**Reply by Phyloiota** [reply]
> > To me the _one person, on vote_ idea only works if we have an on-tangle Digital Identity connected to your real world identity in a unique and ungameable way.
> > I don’t see that happening anytime soon, and even then the discussion of what constitutes “community” is still open. Are people hired on fiver part of it? Is my family? Is my extended family, if I set it up for them? But it’s not just excluding certain votes. Can companies vote? Or only the people in it? Can everyone in the company vote, just because they’re interested? Or because they get cake from the one person who is shilling IOTA? Or what if a competing crypto community suddenly decides they want to vote. Because they hate IOTA’s guts? Or just because Boaty McBoatface?
> > To me ID solutions are out, at least until SSI on IOTA is mature, and maybe not even then. Which I find sad, but I can see it becoming a fair vote anytime soon. Both technically and narrative wise.
> 
> I agree with you about the "personal ID" problem, but to count the votes need a way to identify the voters and validate their options. So the problem here is philosophical: the community is really voting or we just need a way to avoid complains about action that will be take later as result of a moral suasion from huge influencers?
> 
> I guess need a real vote, so we need to grant the voter freedom to vote and to partecipate to the community.
> 
> If this is the magic rule, there is no a real side effect if someone decide to activate a node just to partecipate to the vote and a minute later turn it down. He is free to become a member of the community and there is no additional requirement other than run the node.
> 
> In IOTA 2.0 we will have an additional constraint: a node can vote if it is an "active" member of the community. Being an active member means own lot of MANA, thus move IOTA.
> 
> I guess there is no time to implement MANA based voting system in the current network (but I hope to be wrong), so the only tool we have is the node id.
> 
> If many will take the opportunity to run a node just to partecipate to this pool, I guess it will be simply a good opportunity for the IOTA community to enlarge quickly and involve new adopters.

haha yeah and a lot of people will learn how to setup a Node - thats also value added!

**Reply by basvansambeek** [reply]
> [...] but to count the votes need a way to identify the voters and validate their options.

Why do we need to identify the voters? That is assuming one person = one vote. I’d like that, but it is still up for debate for both practical and philosophical reasons.

> So the problem here is philosophical: the community is really voting or we just need a way to avoid complains about action that will be take later as result of a moral suasion from huge influencers?

I think philosophically a rigged voting process is worse than one where you disagree on a fair voting weight. (It’s been a while ago since I’ve read the research on this, but it seems there is no universal cultural definition of fairness. There is however a general idea of what is unfair across the globe. I can see if I can retrieve the paper. I’m on my phone now.)

In this light we might be better off figuring out which option is deemed most unfair. I’m leaning towards gameable node IDs being that, but perhaps the idea that it teaches people to set up a node could help alleviate it (although with all the one-click solutions available, it might be less attractive again.)

**Reply by sdellava** [reply]
Well, we need to identify the voters simply because we need to count them. Let say that at the end of the pool we have 10.000 votes, it is a correct number? 

Clearly if we decide to use the node id as voter identifier and we have more than 10.000 node, the number of vote appear to be valid, but if we have only 1.000 running nodes, there is a problem.

**Reply by basvansambeek** [reply]
Sorry, I think I misread. I thought you meant human DIDs, but I think you meant machine DIDs. Am I correct there? I’m still getting used to machines having IDs

**Reply by sdellava** [reply]
> Sorry, I think I misread. I thought you meant human DIDs, but I think you meant machine DIDs. Am I correct there? I’m still getting used to machines having IDs

Yes, the node create his own DID (based on the same private key from where his node ID derives) so it is a "Machine DID"

**Reply by rophiroth** [reply]
> > To me the _one person, on vote_ idea only works if we have an on-tangle Digital Identity connected to your real world identity in a unique and ungameable way.
> > I don’t see that happening anytime soon, and even then the discussion of what constitutes “community” is still open. Are people hired on fiver part of it? Is my family? Is my extended family, if I set it up for them? But it’s not just excluding certain votes. Can companies vote? Or only the people in it? Can everyone in the company vote, just because they’re interested? Or because they get cake from the one person who is shilling IOTA? Or what if a competing crypto community suddenly decides they want to vote. Because they hate IOTA’s guts? Or just because Boaty McBoatface?
> > To me ID solutions are out, at least until SSI on IOTA is mature, and maybe not even then. Which I find sad, but I can see it becoming a fair vote anytime soon. Both technically and narrative wise.
> 
> I agree with you about the "personal ID" problem, but to count the votes need a way to identify the voters and validate their options. So the problem here is philosophical: the community is really voting or we just need a way to avoid complains about action that will be take later as result of a moral suasion from huge influencers?
> 
> I guess need a real vote, so we need to grant the voter freedom to vote and to partecipate to the community.
> 
> If this is the magic rule, there is no a real side effect if someone decide to activate a node just to partecipate to the vote and a minute later turn it down. He is free to become a member of the community and there is no additional requirement other than run the node.
> 
> In IOTA 2.0 we will have an additional constraint: a node can vote if it is an "active" member of the community. Being an active member means own lot of MANA, thus move IOTA.
> 
> I guess there is no time to implement MANA based voting system in the current network (but I hope to be wrong), so the only tool we have is the node id.
> 
> If many will take the opportunity to run a node just to partecipate to this pool, I guess it will be simply a good opportunity for the IOTA community to enlarge quickly and involve new adopters.

Maybe instead of constraint could be some kind of "validation grade" and more real info the user have, the more important and deep decisions he is able to vote.

**Comment by dudeamir** [comment]
Something parallel to the whole voting mechanism that would be really cool for future dapps building on top of IOTA to be able to use the same mechanism with minimum effort. If the voting based on token holding for a dapp is desired, would be great to think in advance in a way of how to plug in to this mechanism with a token on IOTA network. 
If lets say nodes are also in the mechanism, maybe a parameter can be set while plugging in to set that weight to zero. 
Just a suggestion!

**Comment by Phyloiota** [comment]
As Dom communicated Yesterday in the Discord - IF will present a solution to put up a vote using the Firefly Wallet.
So they are working on a technical way to casy a "weighted vote" and will put it up and present it to the community.

Thats a fair point and also i like that because i trust it to be probably the most technical advanced option without the possibility to cheat votes if its designed by the IF devs and checked bt the community devs.
So lets assume we will have a working solution to vote based on token holdings. Should we totally rely on this and leave it so for the first round and hope we get a broad participation and  strong decision?
Or should we think about a second layer for node operators?
If we could have a plugin for Hornet thta makes a vote possible and takes Node ID and uptime of the node into account 
we could avoid multiacounting of nodes pretty well i tink:
![image](https://user-images.githubusercontent.com/77154511/117895589-3f12ef00-b2f1-11eb-96cf-dbd5845592ff.png)
So lets say if we do a vote on 30. June, only nodes with an uptime of more than 45 days would be allowed to vote? 
So this would more or less be the nodes that started chrysalis from the beginning onwards before the voting discussion started...

Would still cause some small troubles with people that changed servers or did setup fresh stuff in between but could be

**Comment by rophiroth** [comment]
> We need to find a way to use the IOTA protocol for a fair and immutable voting system. What will be technically doable with the current protocol or future implementations?

What means "inmutable"? not even the voter can change?

**Reply by Phyloiota** [reply]
> > We need to find a way to use the IOTA protocol for a fair and immutable voting system. What will be technically doable with the current protocol or future implementations?
> 
> What means "inmutable"? not even the voter can change?

Yeah - not even the voter should be able to change a vote. So if you give a voting timeframe of 24 hours, or 3 days or whatever. If you have placed a vote once it is fixed. Why should you allow to change your vote every few hours? You have a opinion, or you dont have. Voting is a process that happens since ages and that' s how it works, you vote and your decision is final and neither you nor anybody else can change that vote. 
Its a transaction. It is finalized once confirmed by the ledger, there is no way to rollback a transaction, so should also not be a way to change a vote. Not for a node operator, not for the IF, not for yourself.
This is one of our basic principles in a DLT and it holds for voting also imho.

**Reply by basvansambeek** [reply]
Allowing the voter to change votes creates an attack vector, while it does not help a well intentioned, informed voter.

**Reply by basvansambeek** [reply]
I think this medium post by Samuel Rufinatscha describes the basic characteristics of a voting system (and a potential implementation) here: https://medium.com/@samuel.rufinatscha/e-voting-on-the-tangle-e293d13552a1
Complicating such a system even more seems a high risk/low reward effort.

**Comment by rophiroth** [comment]
Totally agreed that same voting has happens since millennia ago... and
because of that we may change things now, because last 2 millennia wasn't
good enough. If votes cannot be changed we are against the world nature,
because thing change... I can trust communism would help my country now..
but if USA block us or start to kill our people (as it has already happened
before) we could change our mind and change the vote.

**Comment by rophiroth** [comment]
It will always exist attack vectors, well intentioned and well informed
voter will keep informing himself and may need to change his mind. Also
attacks dont have much importance if we can always fix back our vote.. and
with blockchain will be more secured than ever, hacking a single vote will
require strong attack and because of users can easily fix their votes
attackers will think twice if they want to spent their computer potential
to doing something absurd.

And more important... Unmutable and mutable, or time limited votes could
coexist depending on the issue.

**Reply by basvansambeek** [reply]
> It will always exist attack vectors

I don’t think that’s a conveying argument to add new potential attack vectors.

One of the biggest benefits of DLT regarding voting is immutability. To go beyond such a core feature and allow a mutation of a vote is not something I’d change lightly. An ability to alter votes after the fact could also jeopardize the trust in the election.

There *might* be an argument for it, but until we’ve categorically ruled out any weaknesses it could bring to the system, I don’t see a huge benefit to considering it as a feature. Unless there’s already a well vetted example where this is possible.

If incomplete information before a vote is a potential problem, I’d say we focus on fixing that instead. (I don’t see hacking the vote should be possible in any way, given that that is the core of the technology we’re using here.)

**Comment by rophiroth** [comment]
Attacks could be of many ways, an efficient attack that can ruin an
unmutable election is voting before the voter as a voter, or changing his
vote before submitting... Even a bug or human error could submit a wrong
vote.

If vote is editable for the voter only, it would be less important to
trying hacking, because is easy to fix by voter, and also represent a lot
of necessary features as a real time calm long term votations. As more
issues to vote, less like it or important is to hack that one vote, that
the user can easily fix.

What is DLT?

What do you mean by unmutable is the core? You can make your
transactions... And no other, that's user mutable, like editable votes no?

Thank you

**Reply by basvansambeek** [reply]
I’m not sure that being able to change a vote back removes the incentive to hack, but regardless, any voting system should be as safe as possible, and not introduce features that could damage the trust in the voting system. That undermines the whole system that is governed by the election.

**Comment by rophiroth** [comment]
Imagine all the effort to hack 1 vote of 1 issue in millions.. just to see
how easy the user receives a notification, discover it and fix it, would
you keep trying hacking, security system will soon block you all your
available ip and at the end, no vote was finally vulnered. Also some
important votes could be immutable and time limited.

If immutable is the only posibility, hacking one vote or error are
irreversible, so hacker would try their best because the possibility of
accomplish their mission.

**Comment by lukastanisic99** [comment]
It seems we'll have to go with 1 iota = 1 vote. It's hard to do 1 person = 1 vote in a distributed system, because identities can easily be created and the only scarce resource of the network are tokens. Same idea behind MANA which is used as a sibyl protection mechanism is tied to the number of tokens. As I understood the IF they have plans for exploring other sibyl protection mechanisms, but for now MANA is their best bet == tokens. As for the voting mechanism we don't need to complicate anything. Chrysalis uses an UTXO based model so just send a transaction from one of your addresses to another with your message in the payload and after the deadline count all UTXOs that contain the votes. If someone moves their fund the UTXO will be spent and won't be counted.

**Comment by rophiroth** [comment]
What did " scarce resources" mean? Thank you

**Reply by lukastanisic99** [reply]
It means that the resource exists in limited supply and that it can't be easily created. In the case of iota tokens they're finite and they can't be created. On the other hand you have identities that anyone can create multiple (unlimited) times.

**Comment by dudeamir** [comment]
Will it be visible to anyone while the process of voting is happening to see the current status? Seeing x% on optionA at that point and y% on option B? 
I guess it would be easy to do it and show it as the voting goes on. It makes sense as well to show the transparency. However, It can also stir some people to change their mind before casting their vote due to different reasons,... . So what do you guys think? Should be realtime results or after closing the election, counting the votes.

**Reply by lukastanisic99** [reply]
If your question is will there be a GUI that displays the results in real time I don't know. Will the voting be public? Definitely, we're using a public ledger after all.  You'll definitely be able to query the votes and everything will be transparent.

**Reply by Phyloiota** [reply]
Yeah - we dont know yet... I would say that ... if we follow accepted standards for votings the intermediate results should not be public until the vote is closing (thats how its done in usual systems and there are definetly good reasons for that practice)
So i think an extra feature that shows the counts live... No.
But due to the public ledger it will be practical not possible to hide the transactions that contain the vote maybe. Depends which kind of  tx we use etc etc. I think i would prefer a non visible count till the end, but maybe just because we are used to this and it is fairer for everyone to be unbiased over the potential outcome...

**Reply by dudeamir** [reply]
Yeah ultimately I guess its part of the mechanism on how the tx will be so that if anyone can make a real time GUI to show it! It doesnt even need to be from official voting system if the tx is possible to trace. 
So, maybe for the first vote in 5 weeks, its already thought of if there is that possibility or not! Afterward, it can be debated if there is a need to change it.

**Reply by Phyloiota** [reply]
> Yeah ultimately I guess its part of the mechanism on how the tx will be so that if anyone can make a real time GUI to show it! It doesnt even need to be from official voting system if the tx is possible to trace.
> So, maybe for the first vote in 5 weeks, its already thought of if there is that possibility or not! Afterward, it can be debated if there is a need to change it.

Yeah the initial vote in 5 weeks wont be able to "hide" - as we will very likely just use normal transactions on the ledger or something not very fancy, so that will be somehow public from beginning on.... later on in the governance (if established) we can think off a lot of different options and the community should decide (vote) how to vote...

**Comment by Phyloiota** [comment]
Just putting this also up here.... one of several ideas flowting around in my head atm....Nothing fixed and finalised and just to give an idea of what COULD be done!
Open for commenting in the doc....
https://docs.google.com/document/d/1x1Hjx9lhXLrUWBFFh6N_y-9M_lKT_XdxL5yFVy580E8/edit#

**Comment by Phyloiota** [comment]
Idea of a voting flow for Firefly:

![image](https://user-images.githubusercontent.com/77154511/124472293-85a43780-ddd0-11eb-828b-c9b9dda97589.png)
